File: data structures and algorithms/cpp/04_013.cpp
#include <string>
#include <memory>

class Document {
public:
    virtual ~Document() = default;
    virtual void open() = 0;
};

class PDFDocument : public Document {
public:
    void open() override {
        // PDF specific logic
    }
};

class DocumentFactory {
public:
    std::unique_ptr<Document> createDocument(
            const std::string& type) {
        if (type == "pdf") {
            return std::make_unique<PDFDocument>();
        }
        // Add other document types
        return nullptr;
    }
};

File: data structures and algorithms/cs/04_013.cs
public interface IDocument
{
    void Open();
}

public class PDFDocument : IDocument
{
    public void Open() { /* PDF specific logic */ }
}

public class DocumentFactory
{
    public IDocument CreateDocument(string type)
    {
        if (type == "pdf")
        {
            return new PDFDocument();
        }
        // Add other document types
        return null;
    }
}
